Refactor LegacyWebArchive::create functions to take options parameter

szewai opened the pull request:
#### fb3440c796c8c78065a6becda04d2ab8e4fb0b03
<pre>
Refactor LegacyWebArchive::create functions to take options parameter
<a href="https://bugs.webkit.org/show_bug.cgi?id=293451">https://bugs.webkit.org/show_bug.cgi?id=293451</a>
<a href="https://rdar.apple.com/problem/151880030">rdar://problem/151880030</a>

Reviewed by Ryosuke Niwa.

The LegacyWebArchive::create functions become bloated as we add new options for LegacyWebArchive creation. This patch
adds a new struct ArchiveOptions and make these functions take an ArchiveOptions parameter, so that when a new option
is added later, we only need to update ArchiveOptions (without updating declaration of all functions).

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createInternal):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchives):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::webArchiveData):
* Source/WebKitLegacy/mac/DOM/WebDOMOperations.mm:
(-[DOMNode webArchive]):
(-[DOMNode webArchiveByFilteringSubframes:]):
(-[DOMRange webArchive]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _writeSelectionWithPasteboardTypes:toPasteboard:cachedAttributedString:]):

Canonical link: <a href="https://commits.webkit.org/295778@main">https://commits.webkit.org/295778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57b24a468fa647a53971527dadd5891ad41809d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16261 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111316 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/56715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34371 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/56715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109122 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/20958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95764 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/60929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13867 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/56154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/90324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114184 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33256 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33621 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91997 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89385 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22780 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34245 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/12058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/28829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33182 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/38594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32928 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36277 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/34526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -506,7 +506,7 @@ RetainPtr<CFDataRef> LegacyWebArchive::createPropertyListRepresentation(const Re
 #endif
 
-RefPtr<LegacyWebArchive> LegacyWebArchive::create(Node& node, NOESCAPE const Function<bool(LocalFrame&)>& frameFilter, const Vector<MarkupExclusionRule>& customMarkupExclusionRules, const String& mainResourceFilePath, bool saveScriptsFromMemoryCache, ShouldArchiveSubframes shouldArchiveSubframes)
+RefPtr<LegacyWebArchive> LegacyWebArchive::create(Node& node, NOESCAPE const Function<bool(LocalFrame&)>& frameFilter, const LegacyWebArchive::ArchiveOptions& options)
```
Surely we don't need ` LegacyWebArchive::` here and can just use `const ArchiveOptions&` directly?

cdumez replied:
Also, I'm wondering if we could take a `ArchiveOptions&&` here so the implementation can just modify the parameter instead of making a copy. Is there really a call site that cannot move the options in?

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -836,7 +836,7 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, bo
 }
 
-RefPtr<LegacyWebArchive> LegacyWebArchive::createFromSelection(LocalFrame* frame, bool saveScriptsFromMemoryCache, ShouldArchiveSubframes shouldArchiveSubframes)
+RefPtr<LegacyWebArchive> LegacyWebArchive::createFromSelection(LocalFrame* frame, const LegacyWebArchive::ArchiveOptions& options)
```
`LegacyWebArchive::ArchiveOptions` -> `ArchiveOptions`

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
```diff
@@ -43,16 +43,30 @@ struct SimpleRange;
+        {
+        }
+        ShouldArchiveSubframes shouldArchiveSubframes;
+        ShouldSaveScriptsFromMemoryCache shouldSaveScriptsFromMemoryCache;
```
Why can't we just use a boolean here?
`options.shouldSaveScriptsFromMemoryCache = false` is not confusing.

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
```diff
@@ -43,16 +43,30 @@ struct SimpleRange;
+            , shouldSaveScriptsFromMemoryCache(ShouldSaveScriptsFromMemoryCache::Yes)
+        {
+        }
+        ShouldArchiveSubframes shouldArchiveSubframes;
```
We may be able to use a bool here. Now that this is no longer a parameter, we may not need to require an enum class here.

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -515,25 +515,25 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(Node& node, NOESCAPE const Fun
 }
 
-RefPtr<LegacyWebArchive> LegacyWebArchive::create(LocalFrame& frame, ShouldArchiveSubframes shouldArchiveSubframes)
+RefPtr<LegacyWebArchive> LegacyWebArchive::create(LocalFrame& frame, const LegacyWebArchive::ArchiveOptions& options)
```
Unnecessary `LegacyWebArchive::`

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -565,7 +565,7 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(LocalFrame& frame, ShouldArchi
 }
 
-RefPtr<LegacyWebArchive> LegacyWebArchive::create(const SimpleRange& range, bool saveScriptsFromMemoryCache, ShouldArchiveSubframes shouldArchiveSubframes)
+RefPtr<LegacyWebArchive> LegacyWebArchive::create(const SimpleRange& range, const LegacyWebArchive::ArchiveOptions& options)
```
Unnecessary `LegacyWebArchive::`

cdumez commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -686,7 +686,7 @@ static UncheckedKeyHashMap<Ref<CSSStyleSheet>, String> addSubresourcesForCSSStyl
 }
 
-RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, bool saveScriptsFromMemoryCache, ShouldArchiveSubframes shouldArchiveSubframes, LocalFrame& frame, Vector<Ref<Node>>&& nodes, NOESCAPE const Function<bool(LocalFrame&)>& frameFilter, const Vector<MarkupExclusionRule>& markupExclusionRules, const String& mainFrameFilePath)
+RefPtr<LegacyWebArchive> LegacyWebArchive::create(const String& markupString, const LegacyWebArchive::ArchiveOptions& options, LocalFrame& frame, Vector<Ref<Node>>&& nodes, NOESCAPE const Function<bool(LocalFrame&)>& frameFilter)
```
Unnecessary `LegacyWebArchive::`

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
```diff
@@ -43,16 +43,30 @@ struct SimpleRange;
+    struct ArchiveOptions {
+        ArchiveOptions()
+            : shouldArchiveSubframes(ShouldArchiveSubframes::Yes)
+            , shouldSaveScriptsFromMemoryCache(ShouldSaveScriptsFromMemoryCache::Yes)
```
Can we initialize these in member variable declarations and get rid of this constructor?

cdumez replied:
That would be nice indeed. If you do this then please ignore my comments about using booleans and keep using enums classes though. Boolean will be confusing when constructing ArchiveOptions.

cdumez commented on Source/WebCore/editing/cocoa/EditorCocoa.mm:
```diff
@@ -102,7 +102,7 @@
 
     if (elementRange) {
-        if (auto archive = LegacyWebArchive::create(*elementRange)) {
+        if (auto archive = LegacyWebArchive::create(*elementRange, { })) {
```
Why not have a default parameter value?

szewai replied:
I tried adding:
`WEBCORE_EXPORT static RefPtr<LegacyWebArchive> create(const SimpleRange&, ArchiveOptions&& = { });`, and got error:
`error: default member initializer for 'shouldSaveScriptsFromMemoryCache' needed within definition of enclosing class 'LegacyWebArchive' outside of member functions`.
which seems to indicate `ArchiveOptions` cannot be initialized in header.

The code below seems to work; I will update the patch to add `create*` functions without `Options` parameter.
`WEBCORE_EXPORT static RefPtr<LegacyWebArchive> create(const SimpleRange&);`
and
```
RefPtr<LegacyWebArchive> LegacyWebArchive::create(const SimpleRange& range)
{
    return LegacyWebArchive::create(range, { });
}
```

szewai committed:
Refactor LegacyWebArchive::create functions to take options parameter
https://bugs.webkit.org/show_bug.cgi?id=293451
rdar://problem/151880030

Reviewed by Ryosuke Niwa.

The LegacyWebArchive::create functions become bloated as we add new options for LegacyWebArchive creation. This patch
adds a new struct ArchiveOptions and make these functions take an ArchiveOptions parameter, so that when a new option
is added later, we only need to update ArchiveOptions (without updating declaration of all functions).

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createInternal):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchives):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::webArchiveData):
* Source/WebKitLegacy/mac/DOM/WebDOMOperations.mm:
(-[DOMNode webArchive]):
(-[DOMNode webArchiveByFilteringSubframes:]):
(-[DOMRange webArchive]):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView _writeSelectionWithPasteboardTypes:toPasteboard:cachedAttributedString:]):

Canonical link: https://commits.webkit.org/295778@main
